blog/content/post/2021-04-18-optiplex_part2.md

```diff
@@ -0,0 +1,249 @@
+notable frameworks to work with: radare2, Ghidra, IDA and Binary Ninja.
+
+![](https://rehex.ninja/img/thumb/decomp-versus.png)
+
+I decided to give Ghidra a try since it is free and also has some good firmware
+utilities for [EFI binaries](https://github.com/al3xtjames/ghidra-firmware-utils)
+that were developed by a student on GSoC. Note the tool was developed by NSA,
+so if you have any doubts when using it, consider other tools.
+
+The Ghidra firmware utilities are very good to start with, since they parse the
+PE header, look for known GUIDs etc. Additionally it provides EDK2 datatypes
+integration to name variables, structures and interfaces. For example:
+
+![](/img/ghidra_overview.png)
+
+AS you can see the window is divided into code listing and the decompiler. As
+you probably guessed already, the decompiler is our main weapon. Thanks to the
+integrated EDK2 datatypes from UEFI specification we already see the
+human-readable code:
+
+```C
+EFI_STATUS __cdecl _ModuleEntryPoint(EFI_HANDLE ImageHandle,EFI_PEI_SERVICES **PeiServices)
+
+{
+  EFI_STATUS EVar1;
+  
+  ec_clear_interrupts_low();
+  EVar1 = (*(*PeiServices)->InstallPpi)(PeiServices,&EFI_PEI_PPI_DESCRIPTOR_fff75ad0);
+  return EVar1;
+}
+```
+
+There is also a script that translates some known structures and names
+automatically. This helps a lot in obtaining some understandable code
+initially. Note must be taken since the script translates the types from UEFI
+DXE modules (which run from system RAM) and does not recognize PEI modules
+(which run from SPI flash directly in cache). So for example I was making a
+stupid mistake and reversing PEI modules with
+`_ModuleEntryPoint(EFI_HANDLE ImageHandle,EFI_SYSTEM_TABLE *SystemTable)` which
+is only true for DXE and UEFI executables. So it is not always good that a
+script does something automatically for you.
+
+Long story short I have been reversing those modules one by one which were
+related to the SCH5545 and EMI. I spent a lot of time on that and I still have
+got zero clues what to do next. This wasn't however wasted time, because I had
+to do it sooner or later.
+
+# Final conclusion
+
+Many days passed without any action on my side. Then, one morning I have been
+enlightened. "What if the board actually starts but does not give any signs of
+life?" and then... Yesss... After the AC loss the Environmental Controller
+might lost its configuration and that also means other functions may not work
+as well. Before the AC loss I did nto have to configure Environment Controller
+and fan control. The fan was working silently. However just after the AC was
+lost, the fan started to spin at full speed when the board was powered on. It
+is possible that the Environmental Controller keeps its configuration as long
+as standby power is supplied. So taking it into consideration and all my
+previous conclusions I have decided to enable the USB dongle debug instead of
+the onboard serial port. Why? Look at this:
```

Comment:
@miczyg1 maybe you can say precisely where below code coming from?

Reply:
Sure